gc/base/MarkingScheme.cpp

```diff
@@ -435,3 +437,61 @@ MM_MarkingScheme::setupIndexableScanner(MM_EnvironmentBase *env, omrobjectptr_t
 	return _delegate.setupIndexableScanner(env, objectPtr, reason, sizeToDo, sizeInElementsToDo, basePtr, flags);
 }
 
+void
+MM_MarkingScheme::markObjectsForRange(MM_EnvironmentBase *env, void *heapBase, void *heapTop)
+{
+
+	if (heapBase == NULL || heapTop == NULL || (heapBase == heapTop))
+	{
+		return;
+	}
+
+	/* We must atomically mark bits in the first and last bytes of the MarkMap belonging to this range.
+	 * TLH.  This is because other threads may be marking the same byte.
+	 * Internal MarkMap slots to this range cannot be written to by other threads. */
+
+	GC_ObjectHeapIteratorAddressOrderedList objectHeapIterator(env->getExtensions(), (omrobjectptr_t)heapBase, (omrobjectptr_t)heapTop, false, false);
+	uintptr_t slotIndexLow = 0;
+	uintptr_t slotIndexHigh = 0;
+	omrobjectptr_t object = NULL;
+	uintptr_t slotIndex = 0;
+	uintptr_t bitMask = 0;
+	uintptr_t markBlock = 0;
+
+	/* Find the first and last slot we are marking in the mark map */
+	_markMap->getSlotIndexAndMask((omrobjectptr_t)heapBase, &slotIndexLow, &bitMask);
+	/* Remove one off of the range of heapTop, since heapTop is a non-inclusive bound */
+	_markMap->getSlotIndexAndMask((omrobjectptr_t) ((uintptr_t) heapTop - 1), &slotIndexHigh, &bitMask);
+
+	object = objectHeapIterator.nextObject();
+	_markMap->getSlotIndexAndMask(object, &slotIndex , &bitMask);
+
+	/* Atomically set the first slot */
+	while ((NULL != object)  && (slotIndex == slotIndexLow)) {
+		markBlock |= bitMask;
+		object = objectHeapIterator.nextObject();
+		_markMap->getSlotIndexAndMask(object, &slotIndex, &bitMask);
+	}
+
+	_markMap->markBlockAtomic(slotIndexLow, markBlock);
+
+	if (slotIndexLow != slotIndexHigh) {
+
+		/* Mark the middle slots */
+		while ((NULL != object) && (slotIndex != slotIndexHigh)) {
+			_markMap->setBit(object);
+			object = objectHeapIterator.nextObject();
+			_markMap->getSlotIndexAndMask(object, &slotIndex, &bitMask);
+		}
+
+		/* Atomically set the last slot */
+		markBlock = 0;
+		while ((NULL != object) && (slotIndex == slotIndexHigh)) {
+			markBlock |= bitMask;
+			object = objectHeapIterator.nextObject();
+			_markMap->getSlotIndexAndMask(object, &slotIndex, &bitMask);
+		}
+		_markMap->markBlockAtomic(slotIndexHigh, markBlock);
```

Comment:
We don't want to do it by walking the heap and individually marking the objects.
This is more precise approach and ok for maintaining object heap map (what is an optional debug tool), but it's slow for SATB marking GC.
We should try the same thing that Metronome does: blindly mark the whole mark map corresponding to the TLH

Besides, it's probably more optimal to do it as pre  step  (rather than post step, what object map does) to TLH allocation, to avoid a race of those objects being marked through regular object graph traversal and being pushed for scanning (SATB indirectly guaranties they are not required to be scanned), what would be a waste of time (although not incorrect)

Reply:
> We should try the same thing that Metronome does: blindly mark the whole mark map corresponding to the TLH

I attempted this and have been running into some issues.  

I replaced markObjectsForRange that’s implemented in this PR with the code below (copied from `preMarkSmallCells` (`MM_SegregatedMarkingScheme`) ).

Reply:
```C++
void
MM_MarkingScheme::markObjectsForRange(MM_EnvironmentBase *env, void *heapBase, void *heapTop)
{
	if (heapBase == NULL || heapTop == NULL || (heapBase == heapTop))
	{
		return;
	}
	uint8_t *objPtrLow = (uint8_t *)heapBase;
	uint8_t *objPtrHigh = (uint8_t *)heapTop - 1;
	uintptr_t slotIndexLow, slotIndexHigh;
	uintptr_t bitMaskLow, bitMaskHigh;
	_markMap->getSlotIndexAndBlockMask((omrobjectptr_t) objPtrLow, &slotIndexLow, &bitMaskLow, false /* high bit block mask for low slot word */);
	_markMap->getSlotIndexAndBlockMask((omrobjectptr_t) objPtrHigh, &slotIndexHigh, &bitMaskHigh, true /* low bit block mask for high slot word */);
	if (slotIndexLow == slotIndexHigh) {
		_markMap->markBlockAtomic(slotIndexLow, bitMaskLow & bitMaskHigh);
	} else {
		_markMap->markBlockAtomic(slotIndexLow, bitMaskLow);
		_markMap->setMarkBlock(slotIndexLow + 1, slotIndexHigh - 1, (uintptr_t)-1);
		_markMap->markBlockAtomic(slotIndexHigh, bitMaskHigh);
	}
}
```

Reply:
This is crashing/asserting in various ways during sweep. Given that the unoptmial markObjectsForRange was working, I think I missing something with batch marking blindly.

Reply:
### 1. ASSERT: 0x19a00    j9mm.107    *   ** ASSERTION FAILED ** at ParallelSweepScheme.cpp:420: ((false && (holeSize < minimumFreeEntrySize)))

```

#15 0x00007ff04fc79b79 in MM_ParallelSweepScheme::performSamplingCalculations (this=0x7ff05003b260, sweepChunk=0x7ff058149000, markMapCurrent=0x7ff04c1bfe60,
    heapSlotFreeCurrent=0x413b9800) at ParallelSweepScheme.cpp:420
#16 0x00007ff04fc7b951 in MM_ParallelSweepScheme::sweepChunk (this=0x7ff05003b260, env=0x7ff05003b8f8, sweepChunk=0x7ff058149000) at ParallelSweepScheme.cpp:523
#17 0x00007ff04fc91959 in MM_ConcurrentSweepScheme::incrementalSweepChunk (this=0x7ff05003b260, env=0x7ff05003b8f8, chunk=0x7ff058149000) at ConcurrentSweepScheme.cpp:824
#18 0x00007ff04fc91cf7 in MM_ConcurrentSweepScheme::sweepPreviousAvailableChunk (this=0x7ff05003b260, env=0x7ff05003b8f8, sweepState=0x7ff050061ce0)
    at ConcurrentSweepScheme.cpp:877
#19 0x00007ff04fc92e61 in MM_ConcurrentSweepScheme::workThreadFindMinimumSizeFreeEntry (this=0x7ff05003b260, envBase=0x7ff05003b8f8, memorySubSpace=0x7ff0500718b0,
    minimumFreeSize=0) at ConcurrentSweepScheme.cpp:1342
#20 0x00007ff04fc8e690 in MM_ConcurrentSweepFindMinimumSizeFreeTask::run (this=0x7ff0572eade0, env=0x7ff05003b8f8) at ConcurrentSweepScheme.cpp:328
#21 0x00007ff04fb93028 in MM_ParallelDispatcher::run (this=0x7ff050039790, env=0x7ff05003b8f8, task=0x7ff0572eade0, newThreadCount=18446744073709551615)
    at ParallelDispatcher.cpp:588
#22 0x00007ff04fc933eb in MM_ConcurrentSweepScheme::sweepForMinimumSize (this=0x7ff05003b260, env=0x7ff05003b8f8, baseMemorySubSpace=0x7ff050071b70,
    allocateDescription=0x0) at ConcurrentSweepScheme.cpp:1508
#23 0x00007ff04fbf9e0d in MM_ParallelGlobalGC::mainThreadSweepStart (this=0x7ff050039a20, env=0x7ff05003b8f8, allocDescription=0x0) at ParallelGlobalGC.cpp:971
#24 0x00007ff04fbf9715 in MM_ParallelGlobalGC::sweep (this=0x7ff050039a20, env=0x7ff05003b8f8, allocDescription=0x0, rebuildMarkBits=false) at ParallelGlobalGC.cpp:865
#25 0x00007ff04fbf84a8 in MM_ParallelGlobalGC::mainThreadGarbageCollect (this=0x7ff050039a20, env=0x7ff05003b8f8, allocDescription=0x0, initMarkMap=false,
    rebuildMarkBits=false) at ParallelGlobalGC.cpp:465
#26 0x00007ff04fc9e5d4 in MM_ConcurrentGC::internalGarbageCollect (this=0x7ff050039a20, env=0x7ff05003b8f8, subSpace=0x7ff050071b70, allocDescription=0x0)
    at ConcurrentGC.cpp:2103
```

Reply:
### 2. SEG FAULT

```
j> 07:04:14 <gc-op id="29" type="classunload" timems="0.014" contextid="17" timestamp="2021-12-21T07:04:13.098">
j> 07:04:14   <classunload-info classloadercandidates="6" classloadersunloaded="0" classesunloaded="0" anonymousclassesunloaded="0" quiescems="0.000" setupms="0.004" scanms="0.010" postms="0.000" />
j> 07:04:14 </gc-op>
j> 07:04:14 Unhandled exception
j> 07:04:14 Type=Segmentation error vmState=0x0002000c
```
```
#14 0x00007f365c2acaab in GC_ObjectModelDelegate::isIndexable (objectPtr=0x413b5c50, this=0x7f365802e4b0) at ../../../gc_glue_java/ObjectModelDelegate.hpp:315
#15 GC_ObjectModelDelegate::getObjectSizeInBytesWithHeader (includeHashCode=true, objectPtr=0x413b5c50, this=0x7f365802e4b0)
    at ../../../gc_glue_java/ObjectModelDelegate.hpp:231
#16 GC_ObjectModelDelegate::getObjectSizeInBytesWithHeader (objectPtr=0x413b5c50, this=0x7f365802e4b0) at ../../../gc_glue_java/ObjectModelDelegate.hpp:289
#17 GC_ObjectModelBase::getSizeInBytesWithHeader (objectPtr=0x413b5c50, this=0x7f365802e4a0) at ./ObjectModelBase.hpp:211
#18 GC_ObjectModelBase::getConsumedSizeInBytesWithHeader (objectPtr=0x413b5c50, this=0x7f365802e4a0) at ./ObjectModelBase.hpp:229
#19 MM_SweepPoolManagerAddressOrderedListBase::calculateTrailingDetails (trailingCandidateSlotCount=11765, trailingCandidate=0x413b5c58, sweepChunk=0x7f36604aa000,
    this=0x7f3658061630) at SweepPoolManagerAddressOrderedListBase.cpp:86
#20 MM_SweepPoolManagerAddressOrderedListBase::addFreeMemory (this=0x7f3658061630, env=0x7f365803b8f8, sweepChunk=0x7f36604aa000, address=0x413b5c58, size=11765)
    at SweepPoolManagerAddressOrderedListBase.cpp:472
#21 0x00007f365c399ab3 in MM_ParallelSweepScheme::sweepChunk (this=0x7f365803b260, env=0x7f365803b8f8, sweepChunk=0x7f36604aa000) at ParallelSweepScheme.cpp:531
#22 0x00007f365c3af959 in MM_ConcurrentSweepScheme::incrementalSweepChunk (this=0x7f365803b260, env=0x7f365803b8f8, chunk=0x7f36604aa000) at ConcurrentSweepScheme.cpp:824
#23 0x00007f365c3afcf7 in MM_ConcurrentSweepScheme::sweepPreviousAvailableChunk (this=0x7f365803b260, env=0x7f365803b8f8, sweepState=0x7f3658061ce0)
    at ConcurrentSweepScheme.cpp:877
#24 0x00007f365c3b0e61 in MM_ConcurrentSweepScheme::workThreadFindMinimumSizeFreeEntry (this=0x7f365803b260, envBase=0x7f365803b8f8, memorySubSpace=0x7f36580718b0,
    minimumFreeSize=88) at ConcurrentSweepScheme.cpp:1342
#25 0x00007f365c3ac690 in MM_ConcurrentSweepFindMinimumSizeFreeTask::run (this=0x7f365f648b90, env=0x7f365803b8f8) at ConcurrentSweepScheme.cpp:328
--Type <RET> for more, q to quit, c to continue without paging--
#26 0x00007f365c2b1028 in MM_ParallelDispatcher::run (this=0x7f3658039790, env=0x7f365803b8f8, task=0x7f365f648b90, newThreadCount=18446744073709551615)
    at ParallelDispatcher.cpp:588
#27 0x00007f365c3b13eb in MM_ConcurrentSweepScheme::sweepForMinimumSize (this=0x7f365803b260, env=0x7f365803b8f8, baseMemorySubSpace=0x7f3658071b70,
    allocateDescription=0x7f365f649b78) at ConcurrentSweepScheme.cpp:1508
#28 0x00007f365c317e0d in MM_ParallelGlobalGC::mainThreadSweepStart (this=0x7f3658039a20, env=0x7f365803b8f8, allocDescription=0x7f365f649b78) at ParallelGlobalGC.cpp:971
#29 0x00007f365c317715 in MM_ParallelGlobalGC::sweep (this=0x7f3658039a20, env=0x7f365803b8f8, allocDescription=0x7f365f649b78, rebuildMarkBits=false)
    at ParallelGlobalGC.cpp:865
#30 0x00007f365c3164a8 in MM_ParallelGlobalGC::mainThreadGarbageCollect (this=0x7f3658039a20, env=0x7f365803b8f8, allocDescription=0x7f365f649b78, initMarkMap=false,
    rebuildMarkBits=false) at ParallelGlobalGC.cpp:465
#31 0x00007f365c3bc5d4 in MM_ConcurrentGC::internalGarbageCollect (this=0x7f3658039a20, env=0x7f365803b8f8, subSpace=0x7f3658071b70, allocDescription=0x7f365f649b78)
    at ConcurrentGC.cpp:2103
#32 0x00007f365c269393 in MM_Collector::garbageCollect
```

Reply:
Another crash
```
p> 20211221-07:14:23 <gdb> #15 0x00007f35cfb8fe0a in GC_ObjectModelDelegate::isIndexable (objectPtr=0x41c96288, this=0x7f35d002e4c0) at ../../../gc_glue_java/ObjectModelDelegate.hpp:315
p> 20211221-07:14:23 <gdb> #16 GC_ObjectModelDelegate::getObjectSizeInBytesWithHeader (includeHashCode=false, objectPtr=0x41c96288, this=0x7f35d002e4c0) at ../../../gc_glue_java/ObjectModelDelegate.hpp:231
p> 20211221-07:14:23 <gdb> #17 GC_ObjectModelDelegate::getObjectSizeInBytesWithHeader (objectPtr=0x41c96288, this=0x7f35d002e4c0) at ../../../gc_glue_java/ObjectModelDelegate.hpp:289
p> 20211221-07:14:23 <gdb> #18 GC_ObjectModelBase::getSizeInBytesWithHeader (objectPtr=0x41c96288, this=0x7f35d002e4b0) at ./ObjectModelBase.hpp:211
p> 20211221-07:14:23 <gdb> #19 GC_ObjectModelBase::getConsumedSizeInBytesWithHeader (objectPtr=0x41c96288, this=0x7f35d002e4b0) at ./ObjectModelBase.hpp:229
p> 20211221-07:14:23 <gdb> #20 MM_SweepPoolManagerAddressOrderedListBase::addFreeMemory (this=0x7f35d00615b0, env=0x7f35a4001cc8, sweepChunk=0x7f35d8036100, address=0x41c96290, size=6019) at SweepPoolManagerAddressOrderedListBase.cpp:479
p> 20211221-07:14:23 <gdb> #21 0x00007f35cfc7bab3 in MM_ParallelSweepScheme::sweepChunk (this=0x7f35d003b270, env=0x7f35a4001cc8, sweepChunk=0x7f35d8036100) at ParallelSweepScheme.cpp:531
p> 20211221-07:14:23 <gdb> #22 0x00007f35cfc7c207 in MM_ParallelSweepScheme::sweepAllChunks (this=0x7f35d003b270, env=0x7f35a4001cc8, totalChunkCount=3) at ParallelSweepScheme.cpp:640
p> 20211221-07:14:23 <gdb> #23 0x00007f35cfc7c73c in MM_ParallelSweepScheme::internalSweep (this=0x7f35d003b270, env=0x7f35a4001cc8) at ParallelSweepScheme.cpp:775
p> 20211221-07:14:23 <gdb> #24 0x00007f35cfc77d07 in MM_ParallelSweepTask::run (this=0x7f3546c2cdd0, env=0x7f35a4001cc8) at ParallelSweepScheme.cpp:71
p> 20211221-07:14:23 <gdb> #25 0x00007f35cfb91be8 in MM_ParallelDispatcher::workerEntryPoint (this=0x7f35d00397a0, env=0x7f35a4001cc8) at ParallelDispatcher.cpp:186
p> 20211221-07:14:23 <gdb> #26 0x00007f35cfb916dc in dispatcher_thread_proc2 (portLib=0x7f35d69d6e60 <j9portLibrary>, info=0x7f35d71d86d0) at ParallelDispatcher.cpp:92
```

Reply:
I'm not too familiar with the sweep code, I've had a look at Metrnome's sweep scheme and tried to compare with what we do for standard collection. Metrnome is a lot different, perhaps it accounts for blind marking, not sure.

Reply:
Also, I should mention that this is still being done during wipeTLH and not premarked, just trying to minimize changes between working code and this. premark will be next step after I have this working.